File: ios/StillMoment/Domain/Models/WaveformAccumulator.swift
//
//  WaveformAccumulator.swift
//  Still Moment
//
//  Domain Model - Waveform Bucket Builder
//

import Foundation

/// Builds a `MeditationWaveform` by streaming PCM sample chunks into a fixed number
/// of buckets.
///
/// The total frame count is known up front (from the audio file length), so each
/// incoming frame is assigned to a bucket by its global position. Each bucket tracks
/// the peak (maximum absolute amplitude) of the frames that fall into it. `finalize()`
/// normalizes all peaks to the global maximum.
///
/// This is a builder, so it is a small mutable type by design. It contains no
/// AVFoundation/UIKit dependencies and is fully unit-testable with synthetic samples.
struct WaveformAccumulator {
    // MARK: Lifecycle

    /// - Parameters:
    ///   - bucketCount: Number of buckets (waveform bars) to produce. Clamped to at least 1.
    ///   - totalFrameCount: Total number of audio frames that will be appended. Used to map
    ///     each frame to its bucket. A value of `0` produces an all-zero waveform.
    init(bucketCount: Int, totalFrameCount: Int) {
        let safeBucketCount = max(1, bucketCount)
        self.bucketCount = safeBucketCount
        self.totalFrameCount = max(0, totalFrameCount)
        self.peaks = [Float](repeating: 0, count: safeBucketCount)
    }

    // MARK: Internal

    /// Feeds a chunk of PCM samples (mono) sequentially. Samples are assigned to buckets
    /// by their running global frame index. Values may be negative; the absolute value is used.
    mutating func append(samples: [Float]) {
        guard self.totalFrameCount > 0 else {
            return
        }

        for sample in samples {
            let bucketIndex = self.bucketIndex(forFrame: self.processedFrameCount)
            let magnitude = abs(sample)
            if magnitude > self.peaks[bucketIndex] {
                self.peaks[bucketIndex] = magnitude
            }
            self.processedFrameCount += 1
        }
    }

    /// Produces the normalized waveform. Peaks are divided by the global maximum so the
    /// loudest bar becomes `1.0`. An all-silent input (global max `0`) yields all zeros
    /// (no division by zero).
    func finalize() -> MeditationWaveform {
        guard let globalMax = peaks.max(), globalMax > 0 else {
            return MeditationWaveform(samples: [Float](repeating: 0, count: self.bucketCount))
        }

        let normalized = self.peaks.map { $0 / globalMax }
        return MeditationWaveform(samples: normalized)
    }

    // MARK: Private

    private let bucketCount: Int
    private let totalFrameCount: Int
    private var peaks: [Float]
    private var processedFrameCount = 0

    /// Maps a global frame index to its bucket, distributing frames evenly across buckets.
    private func bucketIndex(forFrame frame: Int) -> Int {
        let rawIndex = frame * self.bucketCount / self.totalFrameCount
        return min(rawIndex, self.bucketCount - 1)
    }
}

File: android/app/src/main/kotlin/com/stillmoment/domain/models/WaveformAccumulator.kt
package com.stillmoment.domain.models

import kotlin.math.abs

/**
 * Builds a [MeditationWaveform] by streaming PCM sample chunks into a fixed number
 * of buckets (shared-107).
 *
 * The total frame count is known up front (estimated from the audio file's duration
 * and sample rate on Android), so each incoming frame is assigned to a bucket by its
 * global position. Each bucket tracks the peak (maximum absolute amplitude) of the
 * frames that fall into it. [finalize] normalizes all peaks against the global maximum.
 *
 * This is a mutable builder by design, with no Android framework dependencies — fully
 * unit-testable with synthetic samples. Mirrors the iOS `WaveformAccumulator`.
 *
 * @param bucketCount Number of buckets (waveform bars) to produce. Clamped to at least 1.
 * @param totalFrameCount Total number of mono frames that will be appended. Used to map
 *   each frame to its bucket. A value of `0` produces an all-zero waveform.
 */
class WaveformAccumulator(
    bucketCount: Int,
    totalFrameCount: Int
) {
    private val bucketCount: Int = maxOf(1, bucketCount)
    private val totalFrameCount: Int = maxOf(0, totalFrameCount)
    private val peaks: FloatArray = FloatArray(this.bucketCount)
    private var processedFrameCount: Int = 0

    /**
     * Feeds a chunk of mono PCM samples sequentially. Samples are assigned to buckets
     * by their running global frame index. Values may be negative; the absolute value
     * is used. A no-op when [totalFrameCount] is `0`.
     */
    fun append(samples: List<Float>) {
        if (totalFrameCount == 0) {
            return
        }
        for (sample in samples) {
            val bucketIndex = bucketIndex(processedFrameCount)
            val magnitude = abs(sample)
            if (magnitude > peaks[bucketIndex]) {
                peaks[bucketIndex] = magnitude
            }
            processedFrameCount++
        }
    }

    /**
     * Produces the normalized waveform. Peaks are divided by the global maximum so the
     * loudest bar becomes `1.0`. An all-silent input (global max `0`) yields all zeros
     * (no division by zero).
     */
    fun finalize(): MeditationWaveform {
        val globalMax = peaks.maxOrNull() ?: 0f
        if (globalMax <= 0f) {
            return MeditationWaveform(List(bucketCount) { 0f })
        }
        return MeditationWaveform(peaks.map { it / globalMax })
    }

    /** Maps a global frame index to its bucket, distributing frames evenly. */
    private fun bucketIndex(frame: Int): Int {
        val rawIndex = frame * bucketCount / totalFrameCount
        return minOf(rawIndex, bucketCount - 1)
    }
}
